### DISCUSSION TITLE
Easy transformation/mapping like data-transfer-object

### DISCUSSION THREAD

**Discussion by phh** [spatie/laravel-data#70] [Q&A]
Hi there.

Coming from your data-transfer-object package I was using this to take data from external API request and turn these into a more strict and validated PHP object.

One if the things it made easy was the `mapFrom` annotation. From the docs example:
```
#[MapFrom('address.city')]
public string $city;
```

Is there any sort of same easy behavior, rather than writing a transformer for every case? The `MapFrom`  and `MapTo` seems smart. But maybe this is where the original data-transfer-object is a better choice for this?

**Comment by rubenvanassche** [comment]
Hi @phh,

At the moment we do not have support for such an Attribute but I would love it, so I will certainly merge a PR that adds this functionality.

**Reply by phh** [reply]
Will see if I get the time to do that and look into the internals of the package.
Thanks for your feedback 🙏 .

**Reply by rubenvanassche** [reply]
Looking forward to it!

**Reply by phh** [reply]
Got some time to make something really basic work:

https://github.com/spatie/laravel-data/compare/main...phh:map-from-to-attributes

Tested that something like this:

```php
class MappedData extends Data
{
    public function __construct(
        #[MapFrom('name')]
        #[MapTo('mappedName')]
        public string $mapped_name,
        public string $name,
    ) {
    }
}
```

Data properties looks like this:
```php
MappedData {
    mapped_name: "My name"
    name: "My name"
}
```

And transforms like this:

```php
[
    "mappedName" => "My name"
    "name" => "My name"
]
```

This is still really basic. I will try to make it support multidimensional arrays and add some tests.

**Reply by rubenvanassche** [reply]
@phh You should also check with @aidan-casey, I know he's working on something similar for this package. He's trying to get the injection of route params working using attributes within the data object. 

But that would also open the gate for functionality like yours with mapping properties since those both need to go through a validator when the data is coming from the request.

**Reply by aidan-casey** [reply]
Hey @phh,

The general approach that I have been going down is to move away from validating a request first thing, but instead resolve everything to an array, then validate that array based on the rules in the class, then build the class from that array.

This would let us have an interface (e.g., `Map`) that has a `getValue` method that is used to map various things (e.g., from another key in the array, from a route parameter, etc.)

Happy to talk about this further and collaborate on the implementation.

**Reply by phh** [reply]
@aidan-casey Hi and thanks!

Have you already started doing something in this direction? Would be cool to talk further about the implementation :)

**Reply by aidan-casey** [reply]
Hey @phh,

I have, though I don't have anything cleaned up enough that I have pushed it to a repository. I have been experimenting with some breaking changes to the core that would offer much more flexibility here, but would need to wait until v2.

Happy to talk them over and see if we can collaborate on it.

**Comment by aspyropoulos** [comment]
Yeah that would be really helpful. Voting for that!

**Comment by mnogokotin** [comment]
It's helpfull for me too
